### PR TITLE
Fix #10615: Notation substitution for Ltac2 terms.

### LINKED ISSUE #10615: ltac2 in notations: notation_subst function not found: ltac2:value
Using ltac2 tactic-in-terms under a notation raises an anomaly. Here is a small example.

~~~
From Ltac2
Require Import Ltac2.

Notation "n .+2" := ltac2:(Control.refine (fun () => constr:(S (S n))))
  (at level 20, only parsing).

Definition four := Eval cbv in 2.+2.
~~~

> Error: Anomaly "notation_subst function not found: ltac2:value." Please report at http://coq.inria.fr/bugs/.

Am I doing something wrong? Is this expected? Is there a workaround.

cc/ @ppedrot

### PR DESCRIPTION
**Pull request by ppedrot**
We implement a new type of "preterms" that represent untyped ASTs, corresponding to glob_expr in the ML implementations. Ltac2 quotations inside notations are provided with such preterms, and have to pretype them in order to do anything of relevance with them.

Fixes #10615.

Still needs documentation though.

### CODE REVIEW

**Comment by ppedrot** [comment]
I have another problem though, which I can probably sidestep with dirty hacks, but I'd like to hear from other people before doing so.

In the current implementation, the following example fails with an anomaly.
```
(* Have fun with context capture *)
Notation "[ x ]" := ltac2:(
  let c () := Constr.pretype x in
  refine constr:(forall n : nat, n = ltac2:(Notations.exact0 true c))
).

(* n is not bound anywhere *)
Goal forall [ n ].
Proof.
reflexivity.
Qed.
```
The reason is not directly linked to Ltac2. Rather, notation substitution seems to have a heuristic to dynamically remove terms that look like they are ill-typed. This is performed there.

https://github.com/coq/coq/blob/b6dcb301a1a34df8e4586e8cde6618e7c895fa08/interp/constrintern.ml#L732-L737

Anyone understands why this is done?

**Comment by maximedenes** [comment]
> Goal forall [ n ].

Is there a type on this line?

**Comment by maximedenes** [comment]
A typo I meant :)

**Comment by ppedrot** [comment]
@maximedenes sorry, typo fixed.

**Comment by ppedrot** [comment]
BTW, the constrintern hack is deliberate, see 8d9b66bed22364f513771e9b1fa3df14a3dab866.

File: user-contrib/Ltac2/tac2core.ml [hunk 1]
```diff
@@ -17,6 +17,28 @@ open Tac2expr
 open Tac2entries.Pltac
 open Proofview.Notations
 
+let constr_flags () =
```
Comment (SkySkimmer) [inline user-contrib/Ltac2/tac2core.ml]: Why's this a function?

Comment (ppedrot) [reply user-contrib/Ltac2/tac2core.ml]: Probably historical reasons, it used to have a field that was depending on a global flag. Note that this snippet has not been introduced by this PR, the code was just moved a bit before.

Comment (ppedrot) [reply user-contrib/Ltac2/tac2core.ml]: Fixed.

**Comment by ppedrot** [comment]
I fixed the anomaly by remembering the set of notation variables to be substituted, but this is not very satisfactory. For the time being, I think it's good enough an implementation. The PR is only missing documentation now.

**Comment by ppedrot** [comment]
Documentation added, modulo CI this should be OK.

File: doc/sphinx/proof-engine/ltac2.rst [hunk 2]
```diff
@@ -563,6 +563,20 @@ for it.
 - `&x` as a Coq constr expression expands to
   `ltac2:(Control.refine (fun () => hyp @x))`.
 
+In the special case where Ltac2 antiquotations appear inside a Coq term
+notation, the notation variables are actually systematically bound in the body
```
Comment (jfehrle) [inline doc/sphinx/proof-engine/ltac2.rst]: maybe remove "actually systematically" - doesn't add anything to the sentence, does it?

File: doc/sphinx/proof-engine/ltac2.rst [hunk 3]
```diff
@@ -563,6 +563,20 @@ for it.
...
+of the tactic expression with type `Ltac2.Init.preterm`. Such a type is meant
+to represent untyped syntactic Coq expressions, which can by typed in the
```
Comment (jfehrle) [inline doc/sphinx/proof-engine/ltac2.rst]: -> "This type represents ..."

File: doc/sphinx/proof-engine/ltac2.rst [hunk 4]
```diff
@@ -563,6 +563,20 @@ for it.
...
+current context using the `Ltac2.Constr.pretype` function.
+
+.. example::
+
+   The following notation is essentially the identity.
```
Comment (jfehrle) [inline doc/sphinx/proof-engine/ltac2.rst]: Remove "essentially" or explain how it differs from the identity.

**Review by jfehrle — COMMENTED** [review]
A couple copy-editing suggestions.

File: doc/sphinx/proof-engine/ltac2.rst (hunk 4 above)
Comment (ppedrot) [reply doc/sphinx/proof-engine/ltac2.rst]: It's not the identity, but I don't want to embark in a description of the internals of the pretyper...

**Comment by ppedrot** [comment]
This is ready and needs an assignee. It should be merged before the branch, it's a fairly important PR for Ltac2.

**Comment by ppedrot** [comment]
CI failures are spurious.

**Review by SkySkimmer — APPROVED** [review]
I can't claim to understand this code, but there's no point waiting for the ltac2 secondary maintainer @none so I'll merge unless someone complains within 24h.

**Pull request merged**